python_modules/dagster/dagster/_core/definitions/decorators/assets_definition_factory.py

```diff
@@ -0,0 +1,52 @@
+from functools import cached_property
+from typing import Mapping, NamedTuple, Tuple
+
+from dagster._core.definitions.asset_key import AssetKey
+from dagster._core.definitions.input import In
+from dagster._core.definitions.output import Out
+
+
+class InMapping(NamedTuple):
+    input_name: str
+    input: In
+
+
+class OutMapping(NamedTuple):
+    output_name: str
+    output: Out
+
+
+class InOutMapper:
+    def __init__(
+        self,
+        in_mappings: Mapping[AssetKey, InMapping],
+        out_mappings: Mapping[AssetKey, OutMapping],
+    ) -> None:
+        self.in_mappings = in_mappings
+        self.out_mappings = out_mappings
+
+    @staticmethod
+    def from_asset_ins_and_asset_outs(
+        asset_ins: Mapping[AssetKey, Tuple[str, In]],
+        asset_outs: Mapping[AssetKey, Tuple[str, Out]],
+    ):
+        in_mappings = {
```

Comment:
Instead of doing this conversion here from `Mapping[AssetKey, Tuple[str, In]]` to `Mapping[AssetKey, InMapping]`, did you consider having `build_asset_ins` return a `Mapping[AssetKey, InMapping]`?

Doesn't seem like these unnamed tuples are doing us any good over there either.

Reply:
Agreed we can shoot to get rid of these named tuples but I am trying to keep these changes as incremental and low-risk as possible.

Reply:
because `InMapping` is a `NamedTuple`, it should work in places that accept `Tuple[str, In]`, and thus require no changes to the call sites

I tested this out over here, and pyright seems to pass: https://github.com/dagster-io/dagster/pull/22248. (The other test suite failures seem to be from weird cancellation issues).

Reply:
Addressed in https://github.com/dagster-io/dagster/pull/22251